execution-engine/src/fs.rs

```diff
@@ -0,0 +1,1057 @@
+        let remain_length = (inode_impl.file_stat.file_size - offset) as usize;
+        let offset = offset as usize;
+        if remain_length <= buf.len() {
+            let mut grow_vec = vec![0; buf.len() - remain_length];
+            inode_impl.raw_file_data.append(&mut grow_vec);
+        }
+        let rst = buf.len();
+        inode_impl.raw_file_data[offset..(offset + rst)].copy_from_slice(&buf);
+        inode_impl.file_stat.file_size = inode_impl.raw_file_data.len() as u64;
+        Ok(rst as Size)
+    }
+
+    /// A rust-style base implementation for `fd_read`. It directly calls `fd_pread` with the
+    /// current `offset` of Fd `fd` and then calls `fd_seek`.
+    pub(crate) fn fd_read(&mut self, fd: Fd, len: usize) -> FileSystemError<Vec<u8>> {
+        self.check_right(&fd, Rights::FD_READ)?;
+        let offset = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?.offset;
+
+        let rst = self.fd_pread(fd, len, offset)?;
+        self.fd_seek(fd, rst.len() as i64, Whence::Current)?;
+        Ok(rst)
+    }
+
+    /// The stub implementation of `fd_readdir`. Return unsupported error `NoSys`.
+    #[inline]
+    pub(crate) fn fd_readdir(&mut self, fd: Fd, _cookie: DirCookie) -> FileSystemError<Vec<DirEnt>> {
+        self.check_right(&fd, Rights::FD_READDIR)?;
+        Err(ErrNo::NoSys)
+    }
+
+    /// Atomically renumbers the `old_fd` to the `new_fd`.  Note that as
+    /// execution engine is single-threaded this is atomic from the WASM program's
+    /// point of view.
+    pub(crate) fn fd_renumber(&mut self, old_fd: Fd, new_fd: Fd) -> FileSystemError<()> {
+        let entry = self.fd_table.get(&old_fd).ok_or(ErrNo::BadF)?.clone();
+        if self.fd_table.get(&new_fd).is_none() {
+            self.fd_table.insert(new_fd, entry);
+            self.fd_table.remove(&old_fd);
+            Ok(())
+        } else {
+            Err(ErrNo::BadF)
+        }
+    }
+
+    /// Change the offset of Fd `fd`.
+    pub(crate) fn fd_seek(
+        &mut self,
+        fd: Fd,
+        delta: FileDelta,
+        whence: Whence,
+    ) -> FileSystemError<FileSize> {
+        self.check_right(&fd, Rights::FD_SEEK)?;
+        let FileTableEntry { inode, offset, .. } = self.fd_table.get(&fd).ok_or(ErrNo::BadF)?;
+        let file_size = self
+            .inode_table
+            .get(inode)
+            .ok_or(ErrNo::NoEnt)?
+            .file_stat
+            .file_size;
+
+        let new_base_offset = match whence {
+            Whence::Current => *offset,
+            Whence::End => file_size,
+            Whence::Start => 0,
+        };
+
+        // NOTE: Ensure the computation does not overflow.
+        let new_offset: FileSize = if delta >= 0 {
+            // It is safe to convert a positive i64 to u64.
+            let t_offset = new_base_offset + (delta.abs() as u64);
+            // Offset is allowed to equal to file size
+            if t_offset > file_size {
+                return Err(ErrNo::SPipe);
```

Comment:
Fun fact: seek is often used to expand files when using a file as a random-access buffer. In POSIX, seeking past the end is legal and will results in zeros if followed by a write (read treats the offset as end-of-file):

``` c
// legal in POSIX
f = open("blah", O_RDWR | O_CREAT | O_TRUNC);
seek(f, 1000, SEEK_START);
size(f) => 0;
write(f, "blah");
size(f) => 1004;
```

This is often used by filesystems to let users create "holes", which is a useless optimization here.

Seeking before the beginning of the file is still illegal.

Reply:
I see. Will update it.